src/vm/hints/memset_utils.rs

```diff
@@ -0,0 +1,396 @@
+use crate::bigint;
+use crate::serde::deserialize_program::ApTracking;
+use crate::types::exec_scope::PyValueType;
+use crate::types::relocatable::MaybeRelocatable;
+use crate::vm::errors::vm_errors::VirtualMachineError;
+use crate::vm::hints::hint_utils::get_address_from_var_name;
+use crate::vm::vm_core::VirtualMachine;
+use num_bigint::BigInt;
+use num_traits::FromPrimitive;
+use num_traits::Signed;
+use std::collections::HashMap;
+
+//  Implements hint:
+//  %{ vm_enter_scope({'n': ids.n}) %}
+pub fn memset_enter_scope(
+    vm: &mut VirtualMachine,
+    ids: HashMap<String, BigInt>,
+    hint_ap_tracking: Option<&ApTracking>,
+) -> Result<(), VirtualMachineError> {
+    let n_addr = get_address_from_var_name("n", ids, vm, hint_ap_tracking)?;
+
+    match vm.memory.get(&n_addr) {
+        Ok(Some(maybe_rel_n)) => {
+            let n = if let MaybeRelocatable::Int(n) = maybe_rel_n {
+                n
+            } else {
+                return Err(VirtualMachineError::ExpectedInteger(n_addr.clone()));
+            };
+            vm.exec_scopes.enter_scope(HashMap::from([(
+                String::from("n"),
+                PyValueType::BigInt(n.clone()),
+            )]));
+
+            Ok(())
+        }
+        _ => Err(VirtualMachineError::FailedToGetIds),
+    }
+}
+
+/* Implements hint:
+%{
+    n -= 1
+    ids.continue_loop = 1 if n > 0 else 0
+%}
+*/
+pub fn memset_continue_loop(
+    vm: &mut VirtualMachine,
+    ids: HashMap<String, BigInt>,
+    hint_ap_tracking: Option<&ApTracking>,
+) -> Result<(), VirtualMachineError> {
+    let continue_loop_addr = get_address_from_var_name("continue_loop", ids, vm, hint_ap_tracking)?;
+
+    // get `n` variable from vm scope
+    let n = match vm.exec_scopes.get_local_variables() {
+        Some(variables) => match variables.get("n") {
+            Some(PyValueType::BigInt(n)) => n,
+            _ => {
+                return Err(VirtualMachineError::VariableNotInScopeError(String::from(
+                    "n",
+                )))
+            }
+        },
+        None => return Err(VirtualMachineError::ScopeError),
```

Comment:
I'd collapse these two to return only `VariableNotInScope` instead.

Reply:
I couldn't collapse these two errors.
Since `get_local_variables` returns an option I always need to use two arms.
My best try was implementing this:
```
    // get `n` variable from vm scope
    let n = match vm.exec_scopes.get_local_variables()?.get("n") {
        Some(PyValueType::BigInt(n)) => n,
        _ => {
            return Err(VirtualMachineError::VariableNotInScopeError(String::from(
                "n",
            )))
        }
    };
```

But I got the following error:
`the `?` operator can only be used on `Result`s, not `Option`s, in a function that returns `Result``

Reply:
`Option`s support the `ok_or` method to convert to a `Result`, where the argument is the error it should report if `None`. Then you can use `?` on the result.

Reply:
:party_wizard
[Done!](https://github.com/lambdaclass/cleopatra_cairo/pull/303/commits/c18877ad20e37e07f60be40bf0d57a0a7167d11b)